docs/architecture/design-proposals/third-party-add-repository.md

```diff
@@ -0,0 +1,96 @@
+# Add a repository and trigger a chart deployment from an external source
+
+## Objective
+
+Enable an external user-interface for chart repositories, such as Harbor, to trigger the addition of a repository in Kubeapps or trigger the deployment of a chart directly into the Kubeapps UI.
+
+## Motivation
+
+Kubeapps focuses presenting a catalog of apps from multiple repositories and enabling easy deployment of those apps. Kubeapps does not (and should not) focus on the management of app repositories nor populating them with apps.
+
+Other projects, such as Harbor, focus specifically on the management of repositories for cloud-native resources such as images and apps. These project don't necessarily focus on the user experience for deploying these apps or images.
+
+Enabling these projects to integrate with Kubeapps enables users to move seamlessly from uploading a new app to a repository through to testing the deployment of that same app in their configured Kubeapps installation. Or from preparing a repository with specific apps to enabling users in Kubeapps to access those apps.
+
+## Goals and non goals
+
+* Enable 3rd party sites to provide an "Add repository to Kubeapps" action which results in the Kubeapps installation scanning and including the repository in its catalog.
+* As a consequence, the 3rd party can then link to directly deploy a chart from a repository in the Kubeapps UI.
+
+## Assumptions
+ - The same OpenID Connect identity provider must be used for both the 3rd party (such as Harbor) and Kubeapps (and by implication, the Kubernetes cluster).
+ - RBAC policies in the cluster enable users to deploy apps and for relevant admins, to create AppRepository custom resource definitions (this is part of the Kubeapps installation process).
+
+## User stories
+
+* As a repository maintainer in Harbor, I want to add my new repository to Kubeapps, so that Kubeapps users can deploy charts from my repository.
+* As a contributor to a Harbor repository, I want to test the installation of a chart after uploading to Harbor, so that I can verify the user installation experience.
+* As a repository maintainer in Harbor, I want to remove my repository from Kubeapps, so that users can no longer install apps from my repository until further notice. 
+
+## Implementation
+
+### Kubeapps
+Kubeapps provides an api endpoint for app repositories:
```

Comment:
I'm assuming we want to prevent external access to the Kubernetes API server (otherwise we could just use it directly instead), is that correct? Could this be as simple as a reverse proxy that only exposes the AppRepository CRD endpoints (e.g. a reverse proxy that to something like `http://kubernetes.default/apis/kubeapps.com/v1/apprepositores`)?

Reply:
> Could this be as simple as a reverse proxy that only exposes the AppRepository CRD endpoints (e.g. a reverse proxy that to something like `http://kubernetes.default/apis/kubeapps.com/v1/apprepositores`)?

Yeah I did think along these lines initially, but decided not to head that way as it is exposing Kubeapps implementation details which may change (ie. Kubeapps doesn't need to use a CRD for app repositories - in fact, see #1178 with Andres' proposal to change it... but whether or not that goes ahead at some point, it's an implementation detail which doesn't need to be exposed).

Harbor (or other repo UI) just know that kubeapps should now include repo x using provided credentials - and shouldn't care how Kubeapps keeps that state - both the repo details and the credentials, I think? WDYT? Happy to update the proposal to include that option if you think it's worthwhile considering more widely.